File: samples/typescript/coordination/group_chat.ts
import { ChatAnthropic } from "@langchain/anthropic";
const llm = new ChatAnthropic({ model: "claude-haiku-4-5-20251001" });

type Participant = (thread: [string, string][]) => Promise<string>;
type Manager = (thread: [string, string][], names: string[]) => string | null;

class GroupChat {
  private thread: [string, string][] = [];
  private participants: Map<string, Participant> = new Map();
  constructor(private manager: Manager, private maxTurns = 10) {}
  add(name: string, agent: Participant): this { this.participants.set(name, agent); return this; }
  async run(opening: string): Promise<[string, string][]> {
    this.thread = [["input", opening]];
    const names = [...this.participants.keys()];
    for (let i = 0; i < this.maxTurns; i++) {
      const speaker = this.manager(this.thread, names);
      if (!speaker) return this.thread;
      const agent = this.participants.get(speaker)!;
      const msg = await agent(this.thread);
      this.thread.push([speaker, msg]);
    }
    return this.thread;
  }
}

function makerCheckerManager(approveToken = "APPROVED"): Manager {
  return (thread, _names) => {
    for (let i = thread.length - 1; i >= 0; i--) {
      const [speaker, text] = thread[i];
      if (speaker === "checker") {
        if (text.includes(approveToken)) return null;
        break;
      }
    }
    const last = thread[thread.length - 1][0];
    return last === "maker" ? "checker" : "maker";
  };
}

async function main() {
  console.log("=== Group Chat — Maker-Checker ===\n");
  const maker: Participant = async (thread) => {
    const priorDrafts = thread.filter(([s]) => s === "maker").length;
    const feedback = [...thread].reverse().find(([s]) => s === "checker")?.[1] ?? "none yet";
    const r = await llm.invoke(`You are a writer. Draft #${priorDrafts + 1}. Feedback: ${feedback}. Write a 2-sentence refund policy draft.`);
    return r.content as string;
  };
  const checker: Participant = async (thread) => {
    const draft = [...thread].reverse().find(([s]) => s === "maker")?.[1] ?? "";
    const r = await llm.invoke(`You are a strict reviewer. Goal: clear refund policy. Draft: ${draft}. If good, start with APPROVED. Otherwise start with REJECTED and give one line of feedback.`);
    return r.content as string;
  };
  const chat = new GroupChat(makerCheckerManager(), 8).add("maker", maker).add("checker", checker);
  const thread = await chat.run("draft a refund policy for a SaaS product");
  for (const [speaker, text] of thread) {
    console.log(`[${speaker.padEnd(8)}] ${(text as string).substring(0, 100)}`);
  }
}

main().catch(console.error);

File: samples/python/coordination/group_chat.py
"""Group Chat Pattern — a chat manager drives an accumulating thread.

Framework-agnostic. The chat manager decides who speaks next; agents react to
the shared thread. Includes the turn-based Maker-Checker specialization.

Run: python3 coordination/group_chat.py
See: patterns/coordination/group-chat.md
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable

from langchain_anthropic import ChatAnthropic

# An agent reads the thread (list of (speaker, text)) and returns its message.
Participant = Callable[[list[tuple[str, str]]], str]
# A manager picks the next speaker name given the thread, or None to stop.
Manager = Callable[[list[tuple[str, str]], list[str]], str | None]

llm = ChatAnthropic(model="claude-haiku-4-5-20251001")


@dataclass
class GroupChat:
    """Coordinates a shared, accumulating conversation thread."""

    manager: Manager
    max_turns: int = 10
    participants: dict[str, Participant] = field(default_factory=dict)
    thread: list[tuple[str, str]] = field(default_factory=list)

    def add(self, name: str, agent: Participant) -> "GroupChat":
        self.participants[name] = agent
        return self

    def run(self, opening: str) -> list[tuple[str, str]]:
        self.thread.append(("input", opening))
        names = list(self.participants)
        for _ in range(self.max_turns):
            speaker = self.manager(self.thread, names)
            if speaker is None:
                return self.thread
            message = self.participants[speaker](self.thread)
            self.thread.append((speaker, message))
        return self.thread


def maker_checker_manager(approve_token: str = "APPROVED") -> Manager:
    """Turn-based: maker proposes, checker evaluates. Stops when checker approves."""

    def manager(thread: list[tuple[str, str]], names: list[str]) -> str | None:
        for speaker, text in reversed(thread):
            if speaker == "checker":
                if approve_token in text:
                    return None
                break
        last_speaker = thread[-1][0]
        return "checker" if last_speaker == "maker" else "maker"

    return manager


def llm_maker(thread: list[tuple[str, str]]) -> str:
    prior = [t for s, t in thread if s == "maker"]
    feedback = next((t for s, t in reversed(thread) if s == "checker"), "")
    prompt = (
        f"You are a writer. Previous drafts: {len(prior)}. "
        f"Feedback: {feedback or 'none yet'}. "
        f"Write a short improved draft (2 sentences) for: {thread[0][1]}"
    )
    return llm.invoke(prompt).content


def llm_checker(thread: list[tuple[str, str]]) -> str:
    last_draft = next((t for s, t in reversed(thread) if s == "maker"), "")
    goal = thread[0][1]
    prompt = (
        f"You are a strict reviewer. Goal: {goal}\nDraft: {last_draft}\n"
        f"If the draft fully addresses the goal, reply starting with APPROVED. "
        f"Otherwise start with REJECTED and give one line of feedback."
    )
    return llm.invoke(prompt).content


def main() -> None:
    print("=== Group Chat — Maker-Checker ===\n")
    chat = (
        GroupChat(manager=maker_checker_manager(), max_turns=8)
        .add("maker", llm_maker)
        .add("checker", llm_checker)
    )
    thread = chat.run("draft a two-sentence refund policy for a SaaS product")
    for speaker, text in thread:
        print(f"[{speaker:8}] {text[:120]}")


if __name__ == "__main__":
    main()
